### DISCUSSION TITLE
Why are embedded posts rate-limited?

### DISCUSSION THREAD

**Discussion by GreenFootballs** [bluesky-social/atproto#2459] [Clarifications and "Why?"] [answered]
I actually don't know for sure that embedded posts are being rate limited, so this is kind of a question.

But I recently implemented a system for users of my site to easily include embedded Bluesky posts in their comments, simply by pasting in a URL. The backend code finds URLs in comments, and if it's a Bluesky post URL it goes out and reads the page's HEAD section to get its oembed link, then does a GET to that link for the embed's HTML code.

This all works fine, but sometimes the embeds simply don't display; all we see is the blockquote that's supposed to be rendered by embed.js. When it fails, there's no error in the console, and no indication of why it doesn't display -- but reloading the same post a little later DOES display the embed.

Any idea what's going on here? It seems to me that embedded posts shouldn't be rate limited at all (if that's what's happening), because the un-rendered blockquotes look pretty bad, and serving single un-hydrated static posts shouldn't be much of a bandwidth issue. But I couldn't find any info on this in the Bluesky docs, hence this question.

**Comment by bnewbold** [comment]
My hypothesis for what is going on here is that the individual users are getting limited when they load hundreds of embeds at the same time. Our API endpoints have pretty high rate-limits, but some limit is always needed. Loading a single page with dozens or more embeds results in separate fetches for each embed, which isn't ideal.

The easiest thing we can probably do is bump up the limits a bit. In the long run, some efficiency to only load the embed when the browser scrolls to view it, or do batch requests, or some other efficiency improvement, would all make sense.

**Comment by GreenFootballs** [comment]
Well, I don't think we actually have any pages with hundreds of embeds, but there are probably dozens.

I should add that my code only gets the HTML blockquote tag once, and saves it in the DB. In other words, the oembed step only happens once, when a comment is posted.

But I don't think the current version of embed.js is flexible enough to do batches or lazy loading, is it?

**Reply by bnewbold** [reply, marked as answer]
To be clear, you aren't doing anything wrong, these are all limitations with our setup.

I can confirm I hit HTTP 429 errors when browsing some comment sections on your site with only a handful of embeds. These show up in the network tab of dev tools.

I just bumped the overall rate-limits. It could take a bit of time for those to roll out, but hopefully will resolve this issue for your use case. Please do report again if you are still seeing problems.

**Comment by GreenFootballs** [comment]
Thanks, Bryan.

**Comment by GreenFootballs** [comment]
I posted another request to bump the rate limit, but after more experimentation I think it was a problem on our end, so I deleted that one.
